### DISCUSSION TITLE
File picker problems

### DISCUSSION THREAD

**Discussion by John-Nagle** [emilk/egui#1766] [Q&A]
Ref: https://github.com/emilk/egui/discussions/987

This came up last year. Checking on progress.

I have a file picker dialog working on Linux (Ubuntu 20.04 LTS) and Windows (Wine 7.0). This uses **rfd** in synchronous (modal) mode. The trouble with modal is that if you accidentally click on the main window while the file dialog is up, you're stuck. The file dialog goes behind the 3D game window running Rend3, and the game window is frozen. It's hard to get it back in front if you're full screen. Alt-tab won't help. Minimizing the game window works, but that's not obvious. So, is there some way to keep the file dialog in front of other windows from the same process?

Non-modal would be nice, but the stack corruption troubles in Discussion 987 suggest that doing this asynchronously or in another thread was not a good idea in 2021. Any new info on that?

(Using egui 0.17.0, because Rend3's interface to egui is pinned to that version. Also, I'm running Rend3' framework with EGUI, and it's compute-bound, endlessly refreshing. Not sure that framework ever stops to do async tasks. I use lots of threads but little async, so a threaded solution would be fine. Although threaded apparently fails in full screen mode on Mac.)

**Comment by zicklag** [comment]
I used the `native_dialog` crate instead of `rfd` because `rfd` kept freezing for some reason. You might want to try that crate, as it fixed any problems I ran into. I'm not sure about it falling behind the game, though.

If you want you can check out how I used it in my NESImg app [here](https://github.com/katharostech/NESImg/blob/91b8b5d5960ec38d008255c83bd3bb93c0849f22/src/gui/util.rs#L13), but the app does a lot you probably don't want to do like watching files and stuff, so it may not be the simplest to read example.

**Reply by John-Nagle** [reply]
Thanks. But will that code work on Mac? It spawns a thread to run the dialog.

From the native_dialog docs:
> Why the program crashed when opening a dialog on macOS?
> 
> The UI framework of macOS (Cocoa) has a limitation that all UI operations must be performed on the main thread.

Cocoa was originally backwards compatible with the original single-process, no threads, no CPU dispatcher MacOS. It apparently still has legacy problems from that era.

**Reply by zicklag** [reply]
Ouch. I wasn't aware of that. I guess it probably wouldn't work on Mac. Bummer. :/

**Reply by John-Nagle** [reply]
You'd think there would just be a file picker written in Rust. The trouble is, today's file pickers are privileged. Applications can be run in sandboxes with limited file access. The platform's official file picker grants access rights to a file. On Linux, the file picker can get you access to files outside the application's sandbox. On Mac, it can (I think) get you access to iCloud files. On Windows, it lets you [get access to files outside the app's directory.](https://docs.microsoft.com/en-us/windows/uwp/files/file-access-permissions). 

So we're stuck with native file pickers. But this is such a basic function that it should Just Work.
